File: 0069-sqrtx/0069-sqrtx.py
class Solution:
    def mySqrt(self, x: int) -> int:
        if x < 2:
            return x
        left, right = 1, x // 2
        res = 0
        while left <= right:
            mid = (left+right)//2
            if mid * mid == x:
                return mid
            if mid * mid < x:
                res = mid
                left = mid + 1
            else:
                right = mid - 1
        return res

File: 0069-sqrtx/0069-sqrtx.java
class Solution {
    public int mySqrt(int x) {
        if (x == 0 || x == 1) {
            return x;
        }

        int left = 0, right = x / 2;

        while (left < right) {
            int mid = left + (right - left + 1) / 2;

            long mid2 = (long) mid * (long) mid;

            if (mid2 == (long) x) {
                return mid;
            } else if (mid2 > (long) x) {
                right = mid - 1;
            } else {
                left = mid;
            }
        }
        return left;
    }
}